File: internal/adapters/registry.go
package adapters

import (
	"fmt"
	"os/exec"
	"sort"

	"github.com/JarenKempton/nav/internal/domain"
)

type Kind string

const (
	Tracker   Kind = "tracker"
	Harness   Kind = "harness"
	Workspace Kind = "workspace"
)

type Descriptor struct {
	Name         string               `json:"name"`
	Kind         Kind                 `json:"kind"`
	Bundled      bool                 `json:"bundled"`
	Available    bool                 `json:"available"`
	Executable   string               `json:"executable,omitempty"`
	Capabilities domain.CapabilitySet `json:"capabilities"`
}

func Builtins() []Descriptor {
	trackers := []string{"jira", "linear", "github", "markdown"}
	harnesses := []string{"t3", "pi", "claude", "codex", "cursor", "opencode", "command"}
	result := make([]Descriptor, 0, len(trackers)+len(harnesses)+1)
	for _, name := range trackers {
		result = append(result, Descriptor{Name: name, Kind: Tracker, Bundled: true, Capabilities: domain.NewCapabilities()})
	}
	for _, name := range harnesses {
		executable := map[string]string{"claude": "claude", "codex": "codex", "cursor": "cursor", "opencode": "opencode", "pi": "pi"}[name]
		available := name == "command"
		if executable != "" {
			_, err := exec.LookPath(executable)
			available = err == nil
		}
		caps := domain.NewCapabilities("prompt_generation")
		if available {
			caps["process_launch"] = true
		}
		result = append(result, Descriptor{Name: name, Kind: Harness, Bundled: true, Available: available, Executable: executable, Capabilities: caps})
	}
	result = append(result, Descriptor{Name: "git", Kind: Workspace, Bundled: true, Available: true, Capabilities: domain.NewCapabilities("workspace_prepare")})
	sort.Slice(result, func(i, j int) bool {
		if result[i].Kind == result[j].Kind {
			return result[i].Name < result[j].Name
		}
		return result[i].Kind < result[j].Kind
	})
	return result
}

func Find(name string) (Descriptor, error) {
	for _, descriptor := range Builtins() {
		if descriptor.Name == name {
			return descriptor, nil
		}
	}
	path, err := exec.LookPath("nav-adapter-" + name)
	if err != nil {
		return Descriptor{}, fmt.Errorf("adapter %q not found", name)
	}
	return Descriptor{Name: name, Executable: path, Available: true}, nil
}

File: src/adapters/registry.ts
import { type CapabilitySet, capabilities } from "../domain/model.ts";
import {
  type CommandToken,
  commandHarnessAvailable,
  commandHarnessCapabilities,
  namedHarnessAvailable,
  namedHarnessCapabilities,
  namedHarnesses,
  namedHarnessNameSchema,
} from "./harnesses/command.ts";

export type AdapterKind = "tracker" | "harness" | "workspace" | "environment";

export interface AdapterDescriptor {
  name: string;
  kind: AdapterKind;
  bundled: boolean;
  available: boolean;
  executable?: string;
  capabilities: CapabilitySet;
}

export interface AdapterDiscoveryPlatform {
  which(executable: string): string | null;
  platform: NodeJS.Platform;
}

export interface AdapterRegistryConfiguration {
  command?: { argv: readonly CommandToken[] };
}

const bunDiscoveryPlatform: AdapterDiscoveryPlatform = {
  which: Bun.which,
  platform: process.platform,
};

export function builtInAdapters(
  platform: AdapterDiscoveryPlatform = bunDiscoveryPlatform,
  configuration: AdapterRegistryConfiguration = {},
): AdapterDescriptor[] {
  const trackers = ["jira", "linear", "github", "markdown"].map<AdapterDescriptor>((name) => ({
    name,
    kind: "tracker",
    bundled: true,
    available: false,
    capabilities: {},
  }));

  const namedCommandHarnesses = Object.entries(namedHarnesses).map<AdapterDescriptor>(
    ([key, profile]) => {
      const name = namedHarnessNameSchema.parse(key);
      const executable = profile.argv[0];
      const available = namedHarnessAvailable(name, platform);
      return {
        name,
        kind: "harness",
        bundled: true,
        available,
        ...(executable ? { executable } : {}),
        capabilities: namedHarnessCapabilities(name, platform),
      };
    },
  );
  const commandArgv = configuration.command?.argv;
  const commandExecutable = commandArgv?.[0];
  const harnesses: AdapterDescriptor[] = [
    ...namedCommandHarnesses,
    {
      name: "command",
      kind: "harness",
      bundled: true,
      available: commandHarnessAvailable(commandArgv, platform),
      ...(commandExecutable ? { executable: commandExecutable } : {}),
      capabilities: commandHarnessCapabilities(commandArgv, platform),
    },
    {
      name: "t3",
      kind: "harness",
      bundled: true,
      available: false,
      capabilities: capabilities(),
    },
  ];

  return [
    ...trackers,
    ...harnesses,
    {
      name: "git",
      kind: "workspace" as const,
      bundled: true,
      available: true,
      capabilities: capabilities("workspace_prepare"),
    },
  ].toSorted((left, right) =>
    left.kind === right.kind
      ? left.name.localeCompare(right.name)
      : left.kind.localeCompare(right.kind),
  );
}

export function findAdapter(name: string): AdapterDescriptor {
  const builtIn = builtInAdapters().find((adapter) => adapter.name === name);
  if (builtIn) return builtIn;
  const executable = Bun.which(`wayfinder-adapter-${name}`);
  if (!executable) throw new Error(`Adapter not found: ${JSON.stringify(name)}`);
  return {
    name,
    kind: "tracker",
    bundled: false,
    available: true,
    executable,
    capabilities: {},
  };
}
